src/NetTopologySuite.IO.GPX/GpxReader.cs

```diff
@@ -132,27 +132,19 @@ public static void Read(XmlReader reader, GpxReaderSettings settings, GpxVisitor
                     break;
                 }
 
-                bool expectingMetadata = true;
+                bool foundMetadata = false;
                 bool expectingExtensions = true;
                 do
                 {
-                    if (expectingMetadata)
-                    {
-                        expectingMetadata = false;
-                        if (reader.Name == "metadata")
-                        {
-                            ReadMetadata(reader, settings, creator, visitor);
-                        }
-                        else
-                        {
-                            visitor.VisitMetadata(new GpxMetadata(creator));
-                        }
-                    }
-
                     switch (reader.Name)
                     {
                         // ideally, it should all be in this order, since the XSD validation
                         // would fail otherwise, but whatever.
+                        case "metadata":
+                            ReadMetadata(reader, settings, creator, visitor);
+                            foundMetadata = true;
+                            break;
```

Comment:
`GpxVisitorBase` [documents](https://github.com/NetTopologySuite/NetTopologySuite.IO.GPX/blob/d30ca36c24d2d705a8385478cb7631df06ae2f9e/src/NetTopologySuite.IO.GPX/GpxVisitorBase.cs#L21-L23) that `VisitMetadata` will be called before any other method, for a given GPX file.

I believe that this is important, since if an exception gets thrown while processing a GPX file, the visitor is guaranteed to have seen either the `creator` value if it's seen  anything at all, which the consumer could perhaps use to generate a better error message than what this library does.

Are you **actually** getting GPX files with `metadata` coming out-of-order?  The sample on #41 doesn't do this.

Reply:
It's also [documented](https://github.com/NetTopologySuite/NetTopologySuite.IO.GPX/blob/d30ca36c24d2d705a8385478cb7631df06ae2f9e/src/NetTopologySuite.IO.GPX/GpxVisitorBase.cs#L17) that metadata will only be visited once per file, whereas multiple instances of `metadata` would fall in here multiple times.

Reply:
No, I haven't seen such a GPX file, but if there was one (and judging by the number of issues I opened on bad formatted GPX files I wouldn't be surprised if I see one), it will cause a hang, which I think is worse than not visiting metadata first...
I changed this part of the code mainly because it didn't look elegant as opposed to the other switch case and I don't want this function to hang no matter what I put in.
In case the metadata tag is first it will be processed first, so logically I didn't changed it too much...
In any case, I'll change it to whatever you think is correct.

Reply:
> It's also documented that metadata will only be visited once per file, whereas multiple instances of metadata would fall in here multiple times.
I can add another if to check `foundMetadata` before `ReadMetadata` to make sure it happens only once.